Adds chapter on the Lagrange approach (#156)
moorepants/learn-multibody-dynamics: lagrange.rst

   p.simplify()
   lhs = p.diff(t) - L.jacobian(q)
   rhs = sm.Matrix([0.0, T_b, T_c])

   M = lhs.transpose().jacobian(q_ddot)
   G = lhs.transpose() - M*q_ddot

   q_ddot_sol = M.solve(rhs - G)

.. Practice problem: add a damping force or a coulomb friction force in the first joint 
.. (the example and this problem are inspired by a talk by A. Ruina, https://www.youtube.com/watch?v=j-wHI764dWU)

The generalized momenta are an invertable function of the generalized speeds. We can therefore replace the
Lagrangian equation by:

.. math::

    \dot{p_r} = \frac{\partial L}{\partial q_r}

.. math::

    \dot{q_r} = \dot{q_r}(\bar{p})  

which are equivalent to the equations obtained using Hamilton's method. Hamiltonian systems and their
extension Port-Hamiltonian system are often used in physics and control theory respectively.

For the system described above, the following derives these equations:

.. jupyer-execute::

   p1, p2, p3 = me.dynamicsymbols('p1, p2, p3')
   p_sym = sm.Matrix([p1, p2, p3])
   J_p_wrt_qdot = p.transpose().jacobian(q_dot)
   p_dot = rhs - L.jacobian(q).transpose()
   q_dot_solve = J_p_wrt_qdot.solve(p_sym)

There are two important realizations:

> .. jupyer-execute::

Author: ```suggestion
.. jupyter-execute::
```

============================================
Equations of Motion with the Lagrange Method
============================================

.. note::

   You can download this example as a Python script:

>    :jupyter-download-script:`generalized-forces` or Jupyter Notebook:

Author: ```suggestion
   :jupyter-download-script:`lagrange` or Jupyter Notebook:
```

>    :jupyter-download-notebook:`generalized-forces`.

Author: ```suggestion
   :jupyter-download-notebook:`lagrange`.
```

   import sympy.physics.mechanics as me
   me.init_vprinting(use_latex='mathjax')

.. container:: invisible

   .. jupyter-execute::

      class ReferenceFrame(me.ReferenceFrame):

          def __init__(self, *args, **kwargs):

              kwargs.pop('latexs', None)

              lab = args[0].lower()
              tex = r'\hat{{{}}}_{}'

              super(ReferenceFrame, self).__init__(*args,
                                                   latexs=(tex.format(lab, 'x'),
                                                           tex.format(lab, 'y'),
                                                           tex.format(lab, 'z')),
                                                   **kwargs)
      me.ReferenceFrame = ReferenceFrame

Learning Objectives
===================

After completing this chapter readers will be able to:

- Derive the Lagrangian for a system of particles and rigid bodies
- Use the Euler-Lagrange equation to derive equations of motions given a Lagrangian
- Use the method of Lagrange multipliers to add constraints to the equations of motions

Introduction
============

This book has already discussed three methods to derive the equations

> of motion of multibody systems: Newton-Euler, Kane's method and the TMT

Author: Link to these places in the text where we introduce these other methods.

method. This chapter will add a third: the Lagrange method, originally 

> developed by Joseph-Louis Lagrange.

Reviewer: Should it not be: this chapter adds a **forth**...?

Inertial forces from kinetic energy
===================================

When learning about Kane's method, we learned that the negated generalized inertial
forces equal the applied forces, see :ref:`Unconstrained Equations of Motion`.
A large part of Kane's method of deriving the equations of motions for a 
system is involved with finding the generalized inertial forces.

As an alternative, the following equation also get the generalized inertial forces of a

> system, now by starting from the `kinetic energy TODO`_ :math:`T(\dot{\bar{q}}, \bar{q})`

Author: It is possible to write this as T(u, q, t) and still retain the ability have kinematical differential equations other than q'=u. I've taught it like this before, but it will require `xreplace()` calls in places.

Author: I see you have u in the equation, so it should be consistent either way.

Author: Switch T to K, since I forced your hand :)

expressed as function of the generalized coordinates :math:`\bar{q}`, and 
their time derivatives.

.. _`kinetic energy TODO`: https://en.wikipedia.org/wiki/Work_in_process

.. math::
  :label: eq-lagrange-inertial

>    -\bar{F}^*_r = \frac{\mathrm{d}}{\mathrm{d}t}\left(\frac{\partial T}{\partial u_r}

Author: ```suggestion
   -\bar{F}^*_r = \frac{\mathrm{d}}{\mathrm{d}t}\left(\frac{\partial T}{\partial u}
```

        \right) - \frac{\partial T}{\partial q_r}

.. warning:: Note the two minus signs in the above equation

.. note::

   In Kane's method, we are free to choose generalized speeds :math:`\bar{u}` that differ from
   the time derivatives of the generalized coordinates, that is :math:`\dot{\bar{q}}`. This
   freedom is not readily there when using the Lagrange method. Therefore we will consistently use

>    :math:`\dot{\bar{q}}` in this chapter.

Author: Then change the notation above. You can use things other than q'=u with Lagrange, so I'd just say by convention u=q' is assumed for Lagrange.

The generalized inertial forces computed in this manner are the same as when following

> the TMT method. This can be shown by carefully matching terms in both formulations, as

Reviewer 2: Yes, it would be the same. I wanted to refer out for a "proof" (by example) that they are the same, and just new this one. 

Have added that the Kane's method leads to the same result.

is done for a system of point-masses in [Vallery2020]_.

> Example: 3D body in outer space

Author: If this is the same as https://moorepants.github.io/learn-multibody-dynamics/eom.html#body-fixed-newton-euler-equations, then maybe reference that and compare to it. I was able to display the equations there (if you use a body fixed expression)

Reviewer 2: Added a reference, and explained the similarities and differences. Also see next comment.

.. jupyter-execute::

>    # Setting up reference frames

Author: I try to make separate jupyter cells in general and then move the comments to actual markdown text.

   psi,theta, phi, x, y, z = me.dynamicsymbols('psi theta phi x y z')
   N = me.ReferenceFrame('N')
   B = me.ReferenceFrame('B')

>    B.orient_body_fixed(N, (psi, theta, phi), 'zxy')

Author: This would make it different than my prior example, as these will be ignorable coordinates for a torque free rigid body. That's probably why your equations are long.

Reviewer 2: Yes, the trouble I run into is that I want to just use u = \dot{q}. 

I don't know a clean method (so not the manual substitution used [ for example here](https://phys.libretexts.org/Bookshelves/Classical_Mechanics/Variational_Principles_in_Classical_Mechanics_(Cline)/13%3A_Rigid-body_Rotation/13.18%3A_Lagrange_equations_of_motion_for_rigid-body_rotation). To get back to the nice Euler-Equations

   # Mass and inertia
   m, Ixx, Iyy, Izz = sm.symbols('M, I_{xx}, I_{yy}, I_{zz}')
   I_B = me.inertia(B, Ixx, Iyy, Izz)

   # Kinematics and kinetic energy

   omega_B = B.ang_vel_in(N)
   r_com = x*N.x + y*N.y + z*N.z

>    v_com = r_com.dt(N)

Author: I have common notation for the posistions and velocities. Consistency would be good.

Reviewer 2: Changed them, and checked for others. Probably I missed some. Will keep doing iterations.

>    T = omega_B.dot(I_B.dot(omega_B))/2 + m*v_com.dot(v_com)/2

Author: I need to choose a variable name for kinetic, potential, and total energy. T and V are common for the first two, but I also use T (or bar{T}) for torque, which may be confusing. What do you think of E_k, E_p, and E_t?

Reviewer 2: Yes, most people doing Lagrangian mechanics don't use external torques I guess. Happy with using E_subscripts. Can you confirm when you stick to those?

   # Euler-Lagrange equation

   t = me.dynamicsymbols._t
   q = sm.Matrix([psi, theta, phi, x, y, z])
   qdot = q.diff(t)
   qddot = qdot.diff(t)
   p = sm.Matrix([T]).jacobian(qdot).transpose()

>    g = -sm.Matrix([T]).jacobian(q).transpose()

Author: This is a bit obtuse for the first statement on forming these.

You need to show some math for this too I think.

g is always used for gravity, so that may not be a good variable name.

need to define "p" mathematically

Why are transposes needed here?

Author: The first time I show Lagrange I usually do the equation more expliclty and one by one:

```
T.diff(q1d).diff(t) - V.diff(q1)
```

This matches the equation form more closely.

Then after than I write the equations with Jacobian mathematics and then Jacobian code.

Reviewer 2: Did the first one. Will do a single example of the second one.

   left_hand_side = p.diff(t) + g

This gives the equations of motion, but the terms, particularly the terms
involving :math:`\ddot{q}_r` are mangled. It is common to extract the system
mass matrix and velocity forces vector like so:

.. jupyter-execute::

   mass_matrix = left_hand_side.jacobian(qddot)

>    dynamic_bias = left_hand_side - mass_matrix*qddot

Author: If you use trigsimp on these you may be able to display them and it may not take too long to compute.

The word "dynamic_bias" has not been introduced and there is no math to connect these lines.

Reviewer 2: I'm showing the mass matrix, with some simplification. That looks ok. Now showing the dynamic bias, but it remains a bit lengthy

Reviewer 2: I also replaced dynamic-bias with g_d, with explanation that there is a slight difference. Might need to have a different variables, just as with F_r.

Conservative Forces
===================

Some applied forces, known as conservative forces `conservative forces`_, can
be expressed using the gradient of a scalar function of the generalized coordinates,
known as the `potential energy`_ :math:`V(\bar{q})`:

.. math::
   :label: eq-potential-energy

   \bar{F}_r = -\frac{\partial V}{\partial q_r}

.. warning:: Note the minus sign in the above equation.

.. _`conservative forces`: https://en.wikipedia.org/wiki/Conservative_force
.. _`potential energy`: https://en.wikipedia.org/wiki/Potential_energy

Some examples of conservative forces are:

> * linearized gravity on the surface of the earth, with potential :math:`m g h(\bar{q})`,

Author: I usually say "uniform gravitational field", I'm not sure what "linearized" means here

* gravity from Newton's universal gravitation, with potential :math:`-G \frac{m_1m_2}{r(\bar{q})}`,

> * a linear spring, with potential :math:`\frac{1}{2}k(l(\bar{q}) - l_0)`.

Author: missing squared?

For conservative forces, it is often convenient to derive the applied forces via 
the potential energy.

> The Lagrange-method

Author: ```suggestion
The Lagrange Method
```

===================

Both the equation for computing the inertial forces from the kinetic energy, and 
the equation for computing the applied forces from a potential energy have a term
in them with the partial derivative with respect to the generalized coordinate. 
Furtermore, the potential energy does not depend on the generalized speeds. 
Therefore, we can derive the resulting (inertial and conservative applied) forces
in one go, by combining the two equations.

> Step 1. Compute the so called Lagrangian :math:`L`, the difference between the 

Author: wikipedia link to lagrangian

kinetic energy and potential energy:

.. math::
   :label: eq-lagrangian

   L = T - V

Step 2. Use the Euler-Lagrange equations (the name for the equation 
:ref:`eq-lagrange-inertial`) to find the equations of motion:

.. math::
   :label: eq-euler-lagrange

   \frac{\mathrm{d}}{\mathrm{d}t}\left(\frac{\partial L}{\partial u_r}

>        \right) - \frac{\partial L}{\partial q_r} = \bar{F}_r,

Author: I like how you use Fr here, nice way to avoid having to show them how to arrive at generalized forces (the method I used was based on work and virtual displacements). Cool!

    
while being careful to include a force either in the applied forces 
:math:`\bar{F}_r`, or in the potential energy :math:`V`, but never

> in both.

Author: This is an important note. Should we introduce some notation that breaks Fr into potential related forces and non-conservative ones?

Example: 2D double pendulum with springs and sliding pointmass

Because further examples include multiple bodies, we introduce two convenience functions to
simplify the code for computing the kinetic energy:

.. jupyter-execute::

   def squarednorm(a):
       return a.dot(a)

   def quadraticform(I, v):

>        return v.dot(I.dot(v))

Author: These don't save much typing. I suggest not doing this. I also don't do things like this elsewhere, hoping that the sympy notation is reasonably close to the math notation.

.. jupyter-execute::

   m, g, kt, kl, l = sm.symbols('m, g, k_t, k_l, l')
   q1, q2, q3 = me.dynamicsymbols('q1, q2, q3')

   N = me.ReferenceFrame('N')
   A = me.ReferenceFrame('A')
   B = me.ReferenceFrame('B')

   A.orient_axis(N, q1, N.z)
   B.orient_axis(A, q2, A.x)

   O = me.Point('O')
   Ao = me.Point('A_O')
   Bo = me.Point('B_O')
   Q = me.Point('Q')

   Ao.set_pos(O, l/2*A.x)
   Bo.set_pos(O, l*A.x)
   Q.set_pos(Bo, q3*B.y)

   O.set_vel(N, 0)

   I = m*l**2/12
   I_A_Ao = I*me.outer(A.y, A.y) + I*me.outer(A.z, A.z)
   I_B_Bo = I*me.outer(B.x, B.x) + I*me.outer(B.z, B.z)

Finally, we setup the Lagrangian and derive the equations of motion:

.. jupyter-execute::

   t = sm.symbols('t')
   q = sm.Matrix([q1, q2, q3])
   qdot = q.diff(t)
   qddot = qdot.diff(t)

   T = m/2*(squarednorm(Ao.vel(N)) + squarednorm(Bo.vel(N)) + squarednorm(Q.vel(N))) + 1/2*(
       quadraticform(I_A_Ao, A.ang_vel_in(N)) + quadraticform(I_B_Bo, A.ang_vel_in(N))
   )
   V = m*g*(Ao.pos_from(O).dot(-N.x) + Bo.pos_from(O).dot(-N.x)) + kt/2*(q1**2) + kt/2*q2**2 + kl/2*q3**2

   L = sm.Matrix([T - V])
   lhs = L.jacobian(qdot).diff(t) - L.jacobian(q)
   M = lhs.transpose().jacobian(qddot)

>    G = lhs.transpose() - M*qddot

Author: These should be M_d and g_d that I've defined before, right? If so let's stick with that notation.

Author: I recommend following the same style of replacing all qddot with zero to get G, then you should end up with some simpler equations that you could probably display.

Reviewer 2: The slight advantage of the method I use is that in principle we can check if lhs is indeed linear in the qddots, by checking what symbols exists.

That said, for consistency, and easy understanding, I have changed it.

includes the conservative forces.

Constrained equations of motion
===============================

When using Kane's method, constraints are handled by dividing the generalized speeds into two sets:
the dependent and independent generalized speeds. Then, the dependent generalized speeds are eliminated 
by solving the (time derivative of the) constraint equation.

In the Lagrange method, the generalized speeds should always mach the generalized coordinates.
Therefore, to handle constraints, the generalized coordinates should be likewise eliminated. This
is not possible for non-holonomic constraint (by definition), and requires to solve often difficult
non-linear equations when considering holonomic constraints. This method of elimination is therefore
not useful within the Lagrange method.

Instead, we generalize the approach in :ref:`Exposing Noncontributing Forces`. We will first ommit the
constraint, and add a constraint force, for which we can specify the direction, but not the magnitude. 
The (second) time derivative of the constraint equation is then added to the equations found with the
Euler-Lagrange equation.

> For a particle of mass :math:`m` and position :math:`\bar{r}_{P/O} = q_1 \hat{n}_x + q_2 \hat{n}_y + q_3\hat{n}_z` on a 

Author: ```suggestion
For a particle of mass :math:`m` and position :math:`\bar{r}^{P/O} = q_1 \hat{n}_x + q_2 \hat{n}_y + q_3\hat{n}_z` on a 
```

slope :math:`q_1 = q_2`, we take the second time derivative of the constraint equation :math:`\ddot{q_1} - \ddot{q_2} = 0`,
and a constraint force :math:`\bar{F} = F\hat{n}_x - F\hat{n}_y`. Using the Lagrangian 
:math:`V = \frac{1}{2}m(\dot{q}_1^2 + \dot{q}_2^2 + \dot{q}_3^2) - mgq_3`, we can then derive:

.. math::
    \begin{array}{r}
    m\ddot{q}_1= \phantom{-}F\\
    m\ddot{q}_2= -F\\
    m\ddot{q}_3 + mg = \phantom{-}0\\
    \ddot{q}_1 - \ddot{q}_2\!\! = \phantom{-}0 
    \end{array}

This can be put in matrix-form, by extracting the unknown acceleration and force magnitude;

.. math::
    \begin{bmatrix} m & 0 & 0 &-1 \\ 0 & m & 0 & 1 \\ 0 & 0 & m & 0 \\ 1 & -1 & 0 & 0\end{bmatrix}
    \begin{bmatrix} \ddot{q}_1 \\ \ddot{q}_2 \\ \ddot{q}_3 \\ F \end{bmatrix} = \begin{bmatrix} 0 \\ 0 \\ -mg \\ 0\end{bmatrix}

> It can be tricky to find the direction of the constraint force from the geometric of the system directly.

Author: ```suggestion
It can be difficult to find the direction of the constraint force from the geometric of the system directly.
```

to avoid saying "trick" twice

There is a trick, called the method of the Lagrange multupliers, to quickly find the correct generalized
forces associated with the constraint forces. 

Given a constraint in the general form

.. math::

>     \sum_r a_r(\bar{q}) \dot{q}_r = 0

Author: i may have some notation for this: eq 207 https://moorepants.github.io/learn-multibody-dynamics/holonomic-eom.html maybe we need some way to connect this with that.

We find the generalized force as:

.. math::

    F_r = \lambda a_r(\bar{q})

Here :math:`\lambda` is a variable encoding the magnitude of the constraint force. It is

> called  the Lagrange multiplier. The same :math:`\lambda`` is used for each :math:`r`, that is, 

Author: ```suggestion
called  the Lagrange multiplier. The same :math:`\lambda` is used for each :math:`r`, that is, 
```

        # Euler-Lagrange equation

        t = me.dynamicsymbols._t
        q = sm.Matrix([psi, theta, phi, x, y, z])
        qdot = q.diff(t)
        qddot = qdot.diff(t)
        p = sm.Matrix([T]).jacobian(qdot).transpose()
        g = -sm.Matrix([T]).jacobian(q).transpose()
        left_hand_side = p.diff(t) + g
        mass_matrix = left_hand_side.jacobian(qddot)
        dynamic_bias = left_hand_side - mass_matrix*qddot

To make these free floating body a rolling wheel, three constraints are needed: the
velocity of the contact point should be zero in :math:`\hat{n}_x`, :math:`\hat{n}_y`
and :math:`\hat{n}_x` direction.

.. jupyter-execute::

    lambda1, lambda2, lambda3 = me.dynamicsymbols('lambda1, lambda2, lambda3') 
    constraint = (v_com + B.ang_vel_in(N).cross(-N.z)).to_matrix(N)
    A = constraint.jacobian(qdot)
    diff_constraint = constraint.diff(t)

This constraint information must then be added to the original equations. To do
so, we make use of a useful fact, which is true for all systems and constraints:

.. jupyter-execute::

    diff_constraint.jacobian(qddot) - A

This allows us to create our equations in a block matrix form:

.. math::

>         \begin{bmatrix} M & A^T \\ A & 0\end{bmatrix}\begin{bmatrix}\ddot{\bar{q}} \\ \lambda \end{bmatrix} = 

Author: let's use M_d and g_d also use 'mathbf{M}_d` for matrices

for instance to check no-slip conditions in case of limited friction.

Lagrange's vs Kane's
====================

The Lagrangian method is the  second method to deriving the equations of motion presented in this book,
after Kane's method. This raises the questions: when should each
method be used.

For constrained systems, Kane's method has the advantage that the equations of motion are given for a set of
independent generalized velocities only. This can give rise to simplified equations, additional insight, and
numerically more efficient simulation.

Furthermore, the connection from Kane's method to vector mechanics, that is, Newton's law's, is clearer, which
can provide additional insight, and make it easier to encorporate non-conservative forces such as friction.

On the other hand, the Lagrange-method results in a set of equations with well understood structures and properties.
These structures and properties are not studied further in these materials, other than the following "learn more" section.
With further study, these aspects can make it easier to generalize results across multiple systems, for example

> when designing control algorithms.

Author: Writing the energy expressions only requires you to write magnitudes of velocities, which avoid having to work out accelerations explicitly. Writing energy expressions may be simpler for some problems than all the steps in Kane's method.

With Kane's you don't need Lagrange multipliers.

This example will also show how to apply motor torques at joints.

.. jupyter-execute::

   t, l, r, T_b, T_c = sm.symbols('t, l, r, T_b, T_c')
   q1, q2, q3 = me.dynamicsymbols('q1, q2, q3')

   N = me.ReferenceFrame('N')
   A = me.ReferenceFrame('A')
   B = me.ReferenceFrame('B')
   C = me.ReferenceFrame('C')

   A.orient_axis(N, q1, N.z)
   B.orient_axis(A, q2, A.z)
   C.orient_axis(B, q3, B.x) 

   g = 1
   rho = 1
   m = rho*l*sm.pi*r**2
   I_xx_or_yy = m/12*(3*r**2 + l**2)
   I_zz= m/2*r**2
   I_A_Ao = me.inertia(A, I_xx_or_yy , I_xx_or_yy, I_zz)
   I_B_Bo = me.inertia(B, I_xx_or_yy , I_xx_or_yy, I_zz)
   I_C_Co = me.inertia(C, I_xx_or_yy , I_xx_or_yy, I_zz)

   O = me.Point('O')
   O.set_vel(N, 0.0)
   Ao = me.Point("A_c")
   Ao.set_pos(O, -0.5*l*A.z)
   Bo = me.Point("B_c")
   Bo.set_pos(Ao, -0.5*l*A.z - 0.5*l*B.z)
   Co = me.Point("C_c")
   Co.set_pos(Bo, -0.5*l*B.z -0.5*l*C.z)

The next step is again to form the Lagrangian and find the equations of motion. As the system has no further constraints, 
the Lagrange multiplier method is not needed. The actuator torques are added to the right hand side of the equation, in

> the same way as active forces are added to Kane's equations. Here the torques are represented by the variables :math:`T_b`

Author: ```suggestion
the same way as generalized active forces are added to Kane's equations. Here the torques are represented by the variables :math:`T_b`
```

.. math::

    \min_{q(t)} \int_{0}^{T} L(t, q, \dot{q})\text{d}t \quad \text{s.t.} q(0) = 0, q(T) = q_T  

Examples of such optimizations are:

* The shortest path problem, where :math:`L = |\dot{q}|`
* The brachistochrone problem, that tries to find the shape of a slope, such that a ball rolling off it
  reaches the bottom in minimal time
* Various optimal control problem, in which the integral over the torque squared plus the position error squared
  should be minimized.

For the functional optimization problem, there is again a necessary condition:

.. math::

    \frac{\text{d}}{\text{d}t}\frac{\partial L}{\partial \dot{q}} - \frac{\partial L}{\partial q}= 0,

which we recognize as the Euler-Lagrange equations.

This means that the laws of nature governing rigid body motions result in motions that minimize the integral of the
Lagrangian.  This is called Hamilton's principle. It turns out that many physical laws take such a form of minimizing
the value of a function. One example is Fermat's principle, which states that light takes the path of minimum time.

The optimization point-of-view of the Lagrange method also gives an interpretation for the Lagrange multipliers. They

> are the same as the Lagrange multipliers used in optimization.

Author: I didn't check these two learn more sections in detail, but my main comment is that we don't really have these extensive extra sections in the other chapters. Right now I can tell the students that anything from the chapters we covered can be on the exam. But adding this material makes that easy statement a bit more complex. So my initial reaction is not to include these extensive learn more sections. There are lots of "appendix" materials that I'd like to add (things that don't fit in the classes learning objectives) and we need a way to add them, but we haven't designed that yet. Thoughts?

Author: One more thought: these could be turned in homework problems, then we don't have anything that is "extra".

Reviewer: Most of the math books I read have a section _Notes_ or similar at the end of each chapter. There generally it is stated in which direction the material of the chapter can be extended, maybe some historical notes, etc. Not required reading, but interesting.

Author: Yes, we've added some to the chapters, but only some basic bullet lists to link out to things. These are a bit elaborate with examples and all, so I'm reluctant to include in the chapter. I've tried to make the chapters mostly the minimal set of things needed for the weekly lessons in the course so that students are too distracted with extra info when trying to solve the homeworks.

Reviewer 2: Yes, I agree with removing them. I'll soon push an updated version.

There, I added a very brief paragraph on "generalized momentum", and added knowing it to the learning goals. This paragraph still mentions Noether, and refers to ignorable coordinates (which were defined in the nonholonomic constraint chapter).

Also moved the reference to variational principles to the introduction.

I plan to use the "falling cat" in the homework.

                                                           tex.format(lab, 'y'),
                                                           tex.format(lab, 'z')),
                                                   **kwargs)
      me.ReferenceFrame = ReferenceFrame

Learning Objectives
===================

After completing this chapter readers will be able to:

- Derive the Lagrangian for a system of particles and rigid bodies
- Use the Euler-Lagrange equation to derive equations of motions given a Lagrangian
- Use the method of Lagrange multipliers to add constraints to the equations of motions
- Know how to determine the generalized momenta of a system.

Introduction
============

This book has already discussed two methods to derive the equations
of motion of multibody systems: Newton-Euler and Kane's method. This
chapter will add a third: the `Lagrange method`_, originally 
developed by Joseph-Louis Lagrange. These materials focus on Engineering
applications for multi-body systems, and therefore build the Lagrange method around
the terms found earlier in Kane's equations. In other textbooks, the Lagrange method

> is often derived from the `Variational principles`, such as virtual work or the principle

Author: ```suggestion
is often derived from the `Variational principles`_, such as virtual work or the principle
```

system, now by starting from the `kinetic energy TODO`_ :math:`T(\dot{\bar{q}}, \bar{q})`
expressed as function of the generalized coordinates :math:`\bar{q}`, and 
their time derivatives.

.. _`kinetic energy TODO`: https://en.wikipedia.org/wiki/Work_in_process

.. math::
  :label: eq-lagrange-inertial

   -\bar{F}^*_r = \frac{\mathrm{d}}{\mathrm{d}t}\left(\frac{\partial T}{\partial \dot{q}_r}
        \right) - \frac{\partial T}{\partial q_r}

.. warning:: Note the two minus signs in the above equation

.. note::

   In Kane's method, it is possible to choose generalized speeds :math:`\bar{u}` that differ from
   the time derivatives of the generalized coordinates :math:`\dot{\bar{q}}`. By convention
   :math:`\bar{u} = \dot{\bar{q}}` is assumed when using the Lagrange method. Therefore, throughout
   this chapter :math:`\dot{\bar{q}}` is used.

The generalized inertial forces computed in this manner are the same as when following
Kane's method, or the TMT method used in the next chapter. This can be shown by carefully
matching terms in these formulations, as is done for a a system of point-masses in [Vallery2020]_.

Example: freely moving 3D body

> This example is largely the same as the example in `Body Fixed Newton-Euler Equations`_. A key difference

Author: ```suggestion
This example is largely the same as the example in :ref:`Body Fixed Newton-Euler Equations`. A key difference
```

   T

Use the kinetic energy to find the generalized inertial forces. Here we start with 
the generalized coordinate :math:`\psi`

.. jupyter-execute::

    psid = psi.diff(t)
    F_psi_s = T.diff(psid).diff(t) - T.diff(psi)

A similar derivation should be made for all generalized coordinates. We could write
a loop, but there there is a method to derive all the equations in one go.
The vector of partial derivatives of a function, that is the gradient, can be created
using the jacobian method. The generalized inertial forces can then be find like this: 

.. jupyter-execute::

   T_as_matrix = sm.Matrix([T])
   Fs_transposed = T_as_matrix.jacobian(qd).diff(t) - T_as_matrix.jacobian(q)
   Fs = Fs_transposed.transpose()
   Fs

While these are correct equations of motion, the terms, particularly the terms
involving :math:`\ddot{q}_r` are mangled. It is common to extract the system
mass matrix :math:`\mathbf{M}_d` and velocity forces vector :math:`\bar{g}_d` like before:

.. jupyter-execute::

   qdd_zerod = {qddr: 0 for qddr in qdd}
   Md = Fs.jacobian(qdd)
   gd = Fs.xreplace(qdd_zerod)
   Md.simplify()
   gd.simplify()
   Md, gd

Conservative Forces
===================

> Some applied forces, known as conservative forces `conservative forces`_, can

Author: ```suggestion
Some applied forces, known as `conservative forces`_, can
```

.. _`potential energy`: https://en.wikipedia.org/wiki/Potential_energy

Some examples of conservative forces are:

* a uniform gravitational field, for example on the surface of the earth, with potential :math:`V = m g h(\bar{q})`,
* gravity from Newton's universal gravitation, with potential :math:`V = -G \frac{m_1m_2}{r(\bar{q})}`,
* a linear spring, with potential :math:`V = \frac{1}{2}k(l(\bar{q}) - l_0)^2`.

For conservative forces, it is often convenient to derive the applied forces via 
the potential energy.

The Lagrange Method
===================

Both the equation for computing the inertial forces from the kinetic energy, and 
the equation for computing the applied forces from a potential energy have a term
in them with the partial derivative with respect to the generalized coordinate. 
Furtermore, the potential energy does not depend on the generalized speeds. 
Therefore, the resulting (inertial and conservative applied) forces can be derived
in one go, by combining the two equations.

Step 1. Compute the so called `Lagrangian`_ :math:`L`, the difference between the 
kinetic energy and potential energy:

.. math::
   :label: eq-lagrangian

>    L = T - V

Author: ```suggestion
   L = K - V
```

.. _`Lagrangian`: https://en.wikipedia.org/wiki/Lagrangian

Step 2. Use the Euler-Lagrange equations (the name for the equation 

> :ref:`eq-lagrange-inertial`) to find the equations of motion:

Author: ```suggestion
:math:numref:`eq-lagrange-inertial`) to find the equations of motion:
```

.. math::
   :label: eq-euler-lagrange

   \frac{\mathrm{d}}{\mathrm{d}t}\left(\frac{\partial L}{\partial u_r}
       \right) - \frac{\partial L}{\partial q_r} = \bar{F}_r,
    
while being careful to include a force either in the applied forces 
:math:`\bar{F}_r`, or in the potential energy :math:`V`, but never
in both.

> Example: Double pendulum with springs and sliding pointmass

Author: header?

   B = me.ReferenceFrame('B')

   A.orient_axis(N, q1, N.z)
   B.orient_axis(A, q2, A.x)

   O = me.Point('O')
   Ao = me.Point('A_O')
   Bo = me.Point('B_O')
   Q = me.Point('Q')

   Ao.set_pos(O, l/2*A.x)
   Bo.set_pos(O, l*A.x)
   Q.set_pos(Bo, q3*B.y)

   O.set_vel(N, 0)

   I = m*l**2/12
   I_A_Ao = I*me.outer(A.y, A.y) + I*me.outer(A.z, A.z)
   I_B_Bo = I*me.outer(B.x, B.x) + I*me.outer(B.z, B.z)

Finally, set up the Lagrangian and derive the equations of motion:

.. jupyter-execute::

   t = sm.symbols('t')
   q = sm.Matrix([q1, q2, q3])
   qd = q.diff(t)
   qdd = qd.diff(t)

   T = m/2*(Ao.vel(N).dot(Ao.vel(N)) + Bo.vel(N).dot(Bo.vel(N)) + Q.vel(N).dot(Q.vel(N))) + 1/2*(
       A.ang_vel_in(N).dot(I_A_Ao.dot(A.ang_vel_in(N))) + A.ang_vel_in(N).dot(I_B_Bo.dot(A.ang_vel_in(N)))
   )
   V = m*g*(Ao.pos_from(O).dot(-N.x) + Bo.pos_from(O).dot(-N.x)) + kt/2*(q1**2) + kt/2*q2**2 + kl/2*q3**2

   L = sm.Matrix([T - V])
   left_hand_side = L.jacobian(qd).diff(t) - L.jacobian(q)

   qdd_zerod = {qddr: 0 for qddr in qdd}
   Md = left_hand_side.jacobian(qdd)
   gd = left_hand_side.xreplace(qdd_zerod)

   me.find_dynamicsymbols(Md), me.find_dynamicsymbols(gd)

> The mass matrix :math:`M` only depends on :math:`\bar{q}`, and :math:`\bar{g}` depends

Author: ```suggestion
The mass matrix :math:`\mathbf{M}_d` only depends on :math:`\bar{q}`, and :math:`\bar{g}_d` depends
```

combines the effects of the velocity force vector and the conservative forces. In this setting, 
:math:`\bar{g}_d` is often called the dynamic bias. 

It is often useful to use a vector of intermediate variables when finding the Euler-Lagrange equations. The variables
are defined as:

.. math::

    p_r = \frac{\partial L}{\partial \dot{q_r}}

The variables are collected in a vector :math:`\bar{p}`. They are called the generalized momenta, 
as they coincide with linear momentum in the
case of a Lagrangian describing a particle. Similar to the situation in the dynamics of particles, there can 
be conservation of generalized momentum. This is the case for the generalized momentum associated with ignorable
coordinates, as defined in `Equations of Motion with Nonholonomic Constraints`_. 

Constrained equations of motion
===============================

When using Kane's method, constraints are handled by dividing the generalized speeds into two sets:
the dependent and independent generalized speeds. Then, the dependent generalized speeds are eliminated 
by solving the (time derivative of the) constraint equation.

> In the Lagrange method, the generalized speeds should always match the generalized coordinates.

Author: Maybe also say that the number of dynamical differential equations always equals the number of kinematical differential equations in Lagrange's method. Or something about the number of final equations. Kane's method take a system with N coordinates and give you N kinematical differential equations, p dynamical differential equations, and M holonomic algebraic constraint equations where N - M = n and n - m = p.

Therefore, to handle constraints, the generalized coordinates should be likewise eliminated. This

> is not possible for non-holonomic constraint (by definition), and requires to solve often difficult

Author: ```suggestion
is not possible for holonomic constraint (by definition), and requires to solve often difficult
```

Reviewer 2: Rewrote the sentence to clarify. It really should be non-holonomic in this part though

non-linear equations when considering holonomic constraints. This method of elimination is therefore
not useful within the Lagrange method.

Instead, constraints are handled using a generalized version of the approach in 

> :ref:`Exposing Noncontributing Forces`. First the constraints are omitted. Then a constraint force is added,

Author: ```suggestion
:ref:`Exposing Noncontributing Forces`. First the constraints are omitted. Then a constraint force is added,
```

with a known direction, but unknown magnitude. Finally, the (second) time derivative of the constraint 
equation is then appended to the equations found with the Euler-Lagrange equation.

For example, consider a particle of mass :math:`m` and position 

> :math:`\bar{r}_{P/O} = q_1 \hat{n}_x + q_2 \hat{n}_y + q_3\hat{n}_z`

Author: ```suggestion
:math:`\bar{r}^{P/O} = q_1 \hat{n}_x + q_2 \hat{n}_y + q_3\hat{n}_z`
```

on a slope :math:`q_1 = q_2`.  The unconstrained Lagrangian is 
:math:`V = \frac{1}{2}m(\dot{q}_1^2 + \dot{q}_2^2 + \dot{q}_3^2) - mgq_3`.
The constraint force is perpendicular to the slope, so is described
as :math:`\bar{F} = F\hat{n}_x - F\hat{n}_y`. The appended equation is
the second time derivative of the constraint equation :math:`\ddot{q_1} - \ddot{q_2} = 0`.
Combining all, gives:

.. math::
    \begin{array}{r}
    m\ddot{q}_1= \phantom{-}F\\
    m\ddot{q}_2= -F\\
    m\ddot{q}_3 + mg = \phantom{-}0\\
    \ddot{q}_1 - \ddot{q}_2\!\! = \phantom{-}0 
    \end{array}

> This can be put in matrix-form, by extracting the unknown acceleration and force magnitude;

Author: ```suggestion
This can be put in matrix-form, by extracting the unknown acceleration and force magnitude:
```

.. math::
    \begin{bmatrix} m & 0 & 0 &-1 \\ 0 & m & 0 & 1 \\ 0 & 0 & m & 0 \\ 1 & -1 & 0 & 0\end{bmatrix}
    \begin{bmatrix} \ddot{q}_1 \\ \ddot{q}_2 \\ \ddot{q}_3 \\ F \end{bmatrix} = \begin{bmatrix} 0 \\ 0 \\ -mg \\ 0\end{bmatrix}

It can be challenging to find the direction of the constraint force from the geometry of the system directly.
There is a trick, called the method of the Lagrange multupliers, to quickly find the correct generalized
forces associated with the constraint forces. 

> Given a constraint in the general form

Author: ```suggestion
Given a motion constraint (time derivative of a configuration constraint or a nonholonomic constraint) in the general form
```

.. math::

    \sum_r a_r(\bar{q}) \dot{q}_r = 0

The generalized force is found as:

.. math::

    F_r = \lambda a_r(\bar{q})

Here :math:`\lambda` is a variable encoding the magnitude of the constraint force. It is

> called  the Lagrange multiplier. The same :math:`\lambda` is used for each :math:`r`, that is, 

Author: link to https://en.wikipedia.org/wiki/Lagrange_multiplier here?

each constraint has a single associated Lagrange multiplier.

Due to how it is constructed, the power produced by the constraint force is always zero, as expected.

.. math::

    P = \sum_r F_r\dot{q}_r = \sum \lambda a_r(\bar{q})\dot{q}_r  = \lambda \sum a_r(\bar{q})\dot{q}_r = \lambda \cdot 0

For example, consider the pointmass to be constrained to move in a bowl :math:`x^2 + y^2 + z^2 -1 = 0`.
Taking the time derivative gives: :math:`a_1 = 2q_1` :math:`a_2 = 2q_2` and :math:`a_3 = 2q_3`.
This results in generalized reaction forces :math:`F_1 = 2\lambda q_1`, :math:`F_2 = 2\lambda q_2` and :math:`F_3 = 2\lambda q_3`.

> Example: turning the freely floating body discussed earlier into a rolling sphere.**

Reviewer 2: ** Was a leftover from making it bold. Now a sub-header

The non-slip condition of the rolling sphere is split into three constraints: the velocity of
the contact point (:math:`G`) is zero in the :math:`\hat{n}_x`, the :math:`\hat{n}_y` and  the :math:`\hat{n}_z`

> direction. These constraints are enforced by contact forces in their respective directions.

Author: We should be explicit about this being a configuration constraint or a nonholonomic constraint. You say the velocity is constrained.

The contact point can be found according by :math:`\bar{r}^{G/C} = -r \hat{n}_z`. By using the velocity
two point theorem, the following constraints are found.

.. math::

    \begin{array}{l}
    \bar{n}_x\cdot ({}^N\bar{v}^C + {}^N\bar{\omega}^B \times (-r\hat{n}_z)) = 0 \\
    \bar{n}_y\cdot ({}^N\bar{v}^C + {}^N\bar{\omega}^B \times (-r\hat{n}_z)) = 0 \\
    \bar{n}_z\cdot ({}^N\bar{v}^C + {}^N\bar{\omega}^B \times (-r\hat{n}_z)) = 0 \\
    \end{array}

These can be used to derive the constraint force and the additional equations using the Lagrange-multiplier
method as shown below. Note that here only the first time derivative of the constraint equation is used, 
again because the second time derivatives of the generalized coordinates appear.

> .. container:: invisible

Author: do a dropdown admonition here in case they need to see the setup

    .. jupyter-execute::

        # Setting up reference frames
        psi,theta, phi, x, y, z = me.dynamicsymbols('psi theta phi x y z')
        N = me.ReferenceFrame('N')
        B = me.ReferenceFrame('B')
        B.orient_body_fixed(N, (psi, theta, phi), 'zxy')

        # Mass and inertia
        m, Ixx, Iyy, Izz = sm.symbols('M, I_{xx}, I_{yy}, I_{zz}')
        I_B = me.inertia(B, Ixx, Iyy, Izz)

        # Kinematics and kinetic energy

        omega_B = B.ang_vel_in(N)
        r_com = x*N.x + y*N.y + z*N.z
        v_com = r_com.dt(N)
        T = omega_B.dot(I_B.dot(omega_B))/2 + m*v_com.dot(v_com)/2

        # Euler-Lagrange equation

        t = me.dynamicsymbols._t
        q = sm.Matrix([psi, theta, phi, x, y, z])
        qd = q.diff(t)
        qdd = qd.diff(t)

        L = sm.Matrix([T])
        left_hand_side = L.jacobian(qd).diff(t) - L.jacobian(q)

        qdd_zerod = {qddr: 0 for qddr in qdd}
        Md = left_hand_side.jacobian(qdd)
        gd = left_hand_side.xreplace(qdd_zerod)

> To make these free floating body a rolling wheel, three constraints are needed: the

Author: ```suggestion
To make this free floating body a rolling wheel, three constraints are needed: the
```

velocity of the contact point should be zero in :math:`\hat{n}_x`, :math:`\hat{n}_y`
and :math:`\hat{n}_x` direction.

.. jupyter-execute::

    lambda1, lambda2, lambda3 = me.dynamicsymbols('lambda1, lambda2, lambda3') 
    constraint = (v_com + B.ang_vel_in(N).cross(-N.z)).to_matrix(N)

>     A = constraint.jacobian(qd)

Author: You use lowercase math a in the text and then a cap A here. Can you make this code `A` match a `$A$` in latex math somewhere?

    diff_constraint = constraint.diff(t)

>     constraint

Author: ```suggestion
    sm.trigsimp(constraint)
```

this works nicely for a compact result

This constraint information must then be added to the original equations. To do
so, we make use of a useful fact. 

.. jupyter-execute::

    diff_constraint.jacobian(qdd) - A

This equality is true for all constraints, as can easily be shown by taking the time
derivative of the constraint equation, using the chain rule.

The combined equations can now be written in a block matrix form:

.. math::

>         \begin{bmatrix} \mathbf{M}_d & \mathbf{A}^T \\ \mathbf{A} & 0\end{bmatrix}\begin{bmatrix}\ddot{\bar{q}} \\ \bar{\lambda} \end{bmatrix} = 

Author: define A mathematically before you use it in the code

        \begin{bmatrix} \bar{F}_r - \bar{g}_d \\ - \frac{\partial \mathbf{A}\dot{\bar{q}}}{\partial \bar{q}}\dot{\bar{q}} \end{bmatrix},

> where :math:`A` is the jacobian of the constraints, as used above,  :math:`\bar{g}` is the dynamic bias, and 

Author: ```suggestion
where :math:`\mathbf{A}` is the jacobian of the constraints, as used above,  :math:`\bar{g}` is the dynamic bias, and 
```

>         \begin{bmatrix} \bar{F}_r - \bar{g}_d \\ - \frac{\partial \mathbf{A}\dot{\bar{q}}}{\partial \bar{q}}\dot{\bar{q}} \end{bmatrix},

Author: is the constraint bias term missing parentheses around the first q'?

Author: I have Jacobian notation in the notation page. Would it be helpful to just the Jacobian notation in this chapter? Then maybe you don't even need A or you can define A as the Jacobian notation of `\dot{f}_h`.

Reviewer 2: 1. No, it is a matrix multiplication. I write it inside the partial, because that way everything stays one or two dimensional. Otherwise we'd have to define how to deal with derivatives of a matrix valued function. 
2. To handle earlier comment, I defined a new $\boldsymbol{M}_{hn}$, similar to what is done for holonomic constraints. I use it here now as well.

the last term on the right hand side can be computed as;

.. jupyter-execute::

>     constraint_bias = diff_constraint - diff_constraint.jacobian(qdd)*qdd

Reviewer 3: Was just reading quickly through the chapter to get some inspiration for the multiple choice questions. However this line seems a bit weird/unexpected to me. If I translate the equation before literally I would expect:
```
- (diff_constraint.jacobian(qdd) * qd).jacobian(q) * qd
```

Reviewer 2: Yes, your way works as well. I think there are three (easy) ways to get the bias:

1. Just compute what it should be (like what you do)
2. Set qddot = 0 from the derivative of the constraint (_I will change it to this, because this method is used in similar situations elsewhere in the book_)
3. Substract the parts involving qddot from the second derivative of the constraint (as I did here, I normally do this because it allows me an easy check to see if my assumption on linearity in qddot is correct).

Author: > Substract the parts involving qddot from the second derivative of the constraint (as I did here, I normally do this because it allows me an easy check to see if my assumption on linearity in qddot is correct).

I think this may not always result in the symbolic cancellations that you hope for without simplification. That's why I do the qdd = 0 method in other places.

We call the block matrix called the extended mass matrix, and the vector on the right hand side the extended dynamic bias. 

The method of the Lagrange multiplier can of course also be used within Kane's method. However,

> it results in a larger system of equations, which is why the elimination approach is often

Reviewer 2: Yes, I used 'larger' as in 'larger number of equations' but that is ambiguous.

preferred there. An exception being scenarios where the constraint force itself is a useful output,

> for instance to check no-slip conditions in case of limited friction.

Reviewer 2: Added equation counts throughout the chapter, and fixed the use of "length" for number of equations.

Lagrange's vs Kane's
====================

> The is book has now thought to alternatives to the Newton-Euler method: Kane's method and Lagrange's method. 

Author: ```suggestion
The is book has now presented two alternatives to the Newton-Euler method: Kane's method and Lagrange's method. 
```

This raises the questions: when should each alternative method be used?

For constrained systems, Kane's method has the advantage that the equations of motion are given for a set of

> independent generalized velocities only. This can give rise to simplified equations, additional insight, and

Author: ```suggestion
independent generalized speeds only. This can give rise to simplified equations, additional insight, and
```

numerically more efficient simulation. This also gives the benefit that Lagrange multipliers are not needed
when solving constrained systems with Kane's method.

Furthermore, the connection from Kane's method to vector mechanics, that is, Newton's law's, is clearer, which
can provide additional insight, and make it easier to encorporate non-conservative forces such as friction.

On the other hand, the Lagrange method only requires energies velocities as input, for which only the velocities
of the bodies are needed. This is can be simpler to derive than the accelerations which are needed for Kane's
method.

Furthermore, the Lagrange method results in a set of equations with well understood structures and properties.
These structures and properties are not studied further in these materials. A starting point for further study

> is `Noether's theorem`, which extends the idea of ignorable coordinates to find conserved quantities like

Author: ```suggestion
is `Noether's theorem`_, which extends the idea of ignorable coordinates to find conserved quantities like
```

   rod :math:`B`. Each degree of freedom is resisted by a linear spring. When
   the generalized coordinates are all zero, the two rods are perpendicular to
   each other.

The first step is to define the relevant variables, constants and frames. This step
is the same as for Kane's method.

.. admonition:: Frames and Bodies Setup
   :class: dropdown

    .. jupyter-execute::

        m, g, kt, kl, l = sm.symbols('m, g, k_t, k_l, l')
        q1, q2, q3 = me.dynamicsymbols('q1, q2, q3')

        N = me.ReferenceFrame('N')
        A = me.ReferenceFrame('A')
        B = me.ReferenceFrame('B')

        A.orient_axis(N, q1, N.z)
        B.orient_axis(A, q2, A.x)

        O = me.Point('O')
        Ao = me.Point('A_O')
        Bo = me.Point('B_O')
        Q = me.Point('Q')

        Ao.set_pos(O, l/2*A.x)
        Bo.set_pos(O, l*A.x)
        Q.set_pos(Bo, q3*B.y)

        O.set_vel(N, 0)

        I = m*l**2/12
        I_A_Ao = I*me.outer(A.y, A.y) + I*me.outer(A.z, A.z)
        I_B_Bo = I*me.outer(B.x, B.x) + I*me.outer(B.z, B.z)

Then, set up the Lagrangian:

.. jupyter-execute::

   t = sm.symbols('t')
   q = sm.Matrix([q1, q2, q3])
   qd = q.diff(t)
   qdd = qd.diff(t)

>    K = m/2*(Ao.vel(N).dot(Ao.vel(N)) + Bo.vel(N).dot(Bo.vel(N)) + Q.vel(N).dot(Q.vel(N))) + 1/2*(

Author: ```suggestion
   K = m/2*(Ao.vel(N).dot(Ao.vel(N)) + Bo.vel(N).dot(Bo.vel(N)) + Q.vel(N).dot(Q.vel(N))) + sm.S(1)/2*(
```

Author: Or move the divide by 2 at the end of the expression.

Reviewer 2: I think the divide by 2 suggestion is a bit simpler. Should we add a hint to the sympy chapter for this as well?

Author: I show it in the sympy chapter already (first warning on the page).

Author: sorry not the first one, but there
